src/domain/login/LoginViewModel.js

```diff
@@ -0,0 +1,125 @@
+        this._loginToken = loginToken;
+        this._sessionContainer = this._createSessionContainer();
+        this._loginOptions = null;
+        this._passwordLoginViewModel = null;
+        this._startSSOLoginViewModel = null;
+        this._completeSSOLoginViewModel = null;
+        this._homeserver = null;
+        this._errorMessage = "";
+        this._start(this._defaultHomeServer);
+    }
+
+    get passwordLoginViewModel() { return this._passwordLoginViewModel; }
+    get startSSOLoginViewModel() { return this._startSSOLoginViewModel; }
+    get completeSSOLoginViewModel(){ return this._completeSSOLoginViewModel; }
+    get defaultHomeServer() { return this._defaultHomeServer; }
+    get errorMessage() { return this._errorMessage; }
+
+    async _start(homeserver) {
+        if (this._loginToken) {
+            this._completeSSOLoginViewModel = this.track(new CompleteSSOLoginViewModel(this.childOptions({loginToken: this._loginToken})));
+            this.emitChange("completeSSOLoginViewModel");
+        }
+        else {
+            this._errorMessage = "";
+            await this.queryLogin(homeserver);
+            if (this._loginOptions) {
+                if (this._loginOptions.sso) { this._showSSOLogin(); }
+                if (this._loginOptions.password) { this._showPasswordLogin(); }
+                if (!this._loginOptions.sso && !this._loginOptions.password) {
+                    this._showError("This homeserver neither supports SSO nor Password based login flows");
+                } 
+            }
+            else {
+                this._showError("Could not query login methods supported by the homeserver");
+            }
+        }
+    }
+
+    _showPasswordLogin() {
+        this._passwordLoginViewModel = this.track(new PasswordLoginViewModel(this.childOptions()));
+        this.emitChange("passwordLoginViewModel");
+    }
+
+    _showSSOLogin() {
+        this._startSSOLoginViewModel = this.track(new StartSSOLoginViewModel(this.childOptions()));
+        this.emitChange("startSSOLoginViewModel");
+    }
+
+    _showError(message) {
+        this._errorMessage = message;
+        this.emitChange("errorMessage");
+    }
+
+    async queryLogin(homeserver) {
+        try {
+            this._loginOptions = await this._sessionContainer.queryLogin(homeserver);
+        }
+        catch (e) {
+            this._loginOptions = null;
+        }
+    }
+
+    _disposeViewModels() {
+        this._startSSOLoginViewModel = this.disposeTracked(this._ssoLoginViewModel);
+        this._passwordLoginViewModel = this.disposeTracked(this._passwordLoginViewModel);
+        this.emitChange("disposeViewModels");
+    }
+
+    updateHomeServer(newHomeserver) {
+        this._homeserver = newHomeserver;
+        this._disposeViewModels();
+        this._start(newHomeserver);
+    }
+
+    childOptions(options = {}) {
+        return {
+            ...super.childOptions(options),
+            ready: sessionContainer => {
```

Comment:
so, as discussed, I think it would be worthwhile to move the SessionLoadViewModel up to avoid too many nested view models and code duplication.

I would pass to the child view models a
`attemptLogin(method: LoginMethod): Promise<LoginFailure>` callback that they can await and get a LoginFailure back. This way:
 - child view models can show login errors on their own UI as they see fit
 - child views can still disable their input while attempting the login
 - child views don't have to deal with the SessionLoadViewModel
 - the login view model can also observe the progress of the login and for example disable the homeserver field

So, concretely, once the child view model calls `attemptLogin`, the login view model passes it to startWithLogin on the session container it already created, and wait for the the loadStatus to go from `Login` to something else. If it's a login failure, return that, otherwise return null (meaning the login succeeded). Furthermore, if the login succeeded, it should create a SessionLoadViewModel with the session container passed in. SessionLoadViewModel can now be just about loading the session, and we don't handle the Login(Failure) in there anymore.
Once login succeeded, and we start loading the session, the login options and homeserver field should also be hidden, with only the session load status visible.

Note that `StartSSOLoginViewModel` won't need `attemptLogin`, but `CompleteSSOLoginViewModel` will. The latter will now also also disappear once the token login succeeded, preventing both the "Finishing your sso login" and the label from the SessionLoadViewModel being shown at the same time.